functions/api/category.js

```diff
@@ -0,0 +1,386 @@
+const logger = require('../middleware/logger');
+const config = require('config');
+const joi = require('@hapi/joi');
+const admin = require('firebase-admin');
+const functions = require('firebase-functions');
+const express = require('express');
+const router = express.Router();
+const cors = require('cors');
+const db = admin.firestore();
+
+/**
+ * @description Route to retireve all categories from firestore
+ * @returns Json object containing all categories
+ */
+router.get("/", async (request, response) => {
+    logger.info("Retrieving all categories from firestore");
+    const categories = {
+        "categories": []
+    }
+    let categoryCollection = db.collection(constants.CATEGORY);
+    let snapshot = await categoryCollection.get()
+    snapshot.forEach(category => {
+        var categoryInfo = {}
+        var categoryData = category.data()
+        categoryInfo[constants.CATEGORY] = category.id,
+        categoryInfo[constants.DESCRIPTION] = categoryData.description,
+        categoryInfo[constants.IS_ACTIVE] = categoryData.isActive,
+        categoryInfo[constants.CREATED_DATE] = categoryData.createdDate.toDate(),
+        categoryInfo[constants.LAST_UPDATED_DATE] = categoryData.lastUpdatedDate.toDate()
+        categories.categories.push(categoryInfo);
+    })
+    categories[constants.TOTAL_CATEGORIES] = snapshot.size;
+    logger.debug('Returning categories to client.');
+    response.status(200).send(categories);
+});
+
+/**
+ * @description Route to retrieve single category data from firestore
+ * @returns Json object containing requested category
+ * @throws 400 if the product does not exists in firestore
+ */
+router.get('/:category', async (request, response, next) => {
+    var  requestedCategory = request.params.category.toLocaleLowerCase()
+    logger.info(`Retrieving category ${requestedCategory} from firestore`)
+    var categoryInfo = {}
+    const doc = db.collection(constants.CATEGORY).doc(requestedCategory);
+    const category = await doc.get()
+    if (!category.exists) {
+        const error = new Error(`Requested category ${requestedCategory} is not present in Firestore.`)
+        error.statusCode = 404
+        next(error)
+        return;
+    }
+    var categoryData = category.data()
+    categoryInfo[constants.CATEGORY] = category.id
+    categoryInfo[constants.DESCRIPTION] = categoryData.description
+    categoryInfo[constants.IS_ACTIVE] = categoryData.isActive
+    categoryInfo[constants.CREATED_DATE] = categoryData.createdDate.toDate()
+    categoryInfo[constants.LAST_UPDATED_DATE] = categoryData.lastUpdatedDate.toDate()
+    logger.debug(`Returning details for category ${requestedCategory} to client.`);
+    response.status(200).send(categoryInfo);
+});
+
+/**
+ * @description Route to retrieve all products from a given category
+ * @returns Json object containing array of products for a given category
+ */
+router.get('/products/:category', async (request, response, next) => {
```

Comment:
shouldnt this api be part of products route?

Reply:
This is kind of redundant. getProducts from in product js returns all info about the products. This will return only array of product names. We may not use this. I was just trying to cover all fields from category

Reply:
so, do we need?

Reply:
I am not sure. But if user wants to add a product, we will give the list of all categories available. With that we can show already existing products under that category. This will reduce one read.